internal/provision/devices.go

```diff
@@ -130,3 +91,109 @@ func LoadDevices(path string, dic *di.Container) errors.EdgeX {
 
 	return nil
 }
+
+func loadDevicesFromFile(path string, serviceName string, lc logger.LoggingClient, addDevicesReq []requests.AddDeviceRequest) errors.EdgeX {
+	absPath, err := filepath.Abs(path)
+	if err != nil {
+		return errors.NewCommonEdgeX(errors.KindServerError, "failed to create absolute path", err)
+	}
+
+	files, err := os.ReadDir(absPath)
+	if err != nil {
+		return errors.NewCommonEdgeX(errors.KindServerError, "failed to read directory", err)
+	}
+
+	if len(files) == 0 {
+		return nil
+	}
+
+	lc.Infof("Loading pre-defined devices from %s(%d files found)", absPath, len(files))
+
+	for _, file := range files {
+		fullPath := filepath.Join(absPath, file.Name())
+		addDevicesReq = processDevices(fullPath, serviceName, nil, lc, addDevicesReq)
+	}
+	return nil
+}
+
+func loadDevicesFromUri(inputUri string, serviceName string, secretProvider interfaces.SecretProvider, lc logger.LoggingClient, addDevicesReq []requests.AddDeviceRequest) errors.EdgeX {
+	parsedUrl, err := url.Parse(inputUri)
+	if err != nil {
+		return errors.NewCommonEdgeX(errors.KindServerError, "could not parse uri for Device", err)
+	}
+
+	bytes, err := file.Load(inputUri, secretProvider, lc)
+	if err != nil {
+		return errors.NewCommonEdgeX(errors.KindServerError, fmt.Sprintf("failed to read uri %s", parsedUrl.Redacted()), err)
+	}
+
+	if len(bytes) == 0 {
+		return nil
+	}
+
+	var files []string
+
+	err = json.Unmarshal(bytes, &files)
+	if err != nil {
+		return errors.NewCommonEdgeX(errors.KindServerError, "could not unmarshal Device contents", err)
+	}
+
+	if len(files) == 0 {
+		return nil
+	}
+
+	baseUrl, _ := path.Split(inputUri)
+	lc.Infof("Loading pre-defined devices from %s(%d files found)", parsedUrl.Redacted(), len(files))
+
+	for _, file := range files {
+		fullPath, err := url.JoinPath(baseUrl, file)
+		if err != nil {
+			lc.Error("could not join uri path for device %s: %v", file, err)
+			continue
+		}
+		addDevicesReq = processDevices(fullPath, serviceName, secretProvider, lc, addDevicesReq)
+	}
+	return nil
+}
+
+func processDevices(fullPath string, serviceName string, secretProvider interfaces.SecretProvider, lc logger.LoggingClient, addDevicesReq []requests.AddDeviceRequest) []requests.AddDeviceRequest {
+	var devices []dtos.Device
+	content, err := file.Load(fullPath, secretProvider, lc)
+	if err != nil {
+		lc.Errorf("Failed to read %s: %v", fullPath, err)
+		return addDevicesReq
+	}
+	if strings.HasSuffix(fullPath, yamlExt) || strings.HasSuffix(fullPath, ymlExt) {
+		d := struct {
+			DeviceList []dtos.Device `yaml:"deviceList"`
+		}{}
+		err = yaml.Unmarshal(content, &d)
+		if err != nil {
+			lc.Errorf("Failed to YAML decode %s: %v", fullPath, err)
+			return addDevicesReq
+		}
+		devices = d.DeviceList
+	} else if strings.HasSuffix(fullPath, ".json") {
+		err = json.Unmarshal(content, &devices)
+		if err != nil {
+			lc.Errorf("Failed to JSON decode %s: %v", fullPath, err)
+			return addDevicesReq
+		}
+	} else {
+		return addDevicesReq
```

Comment:
Put comment `why` skipping and not an error.
Also this check if file is YAML or JSON should go before file.Load() so we can return early if no YAML or JSON.
I think a helper function that return the file type would be good here since it can be used for the Profiles and Provision Watchers. Use constants for the returned type  (YAML, JSON, OTHER)